bokeh/charts/_builder.py

```diff
@@ -0,0 +1,321 @@
+        return self.draw()
+
+
+    def create(self, chart=None):
+        self.pre_create()
+
+        # pass these renderers to the chart and then forget about the chart
+        renderers = self.make_renderers()
+        chart.add_renderers(self, renderers)
+
+        # create chart ranges..
+        if not chart.x_range:
+            chart.x_range = self.x_range
+        if not chart.y_range:
+            chart.y_range = self.y_range
+
+        # create the legends if needed
+        if self._legend:
+            if self._legend is True:
+                    orientation = "top_right"
+            else:
+                orientation = self._legend
+
+            legends = self._legends
+            chart.add_legend(orientation, legends)
+
+        return chart
+
+    #***************************
+    # Some helper methods
+    #***************************
+    def _chunker(self, l, n):
+        """Yield successive n-sized chunks from l.
+
+        Args:
+            l (list: the incomming list to be chunked
+            n (int): lenght of you chucks
+        """
+        for i in range(0, len(l), n):
+            yield l[i:i + n]
+
+    def _set_colors(self, chunk):
+        """Build a color list just cycling through a defined palette.
+
+        Args:
+            chuck (list): the chunk of elements to generate the color list.
+        """
+        colors = []
+
+        pal = ["#f22c40", "#5ab738", "#407ee7", "#df5320", "#00ad9c", "#c33ff3"]
+        import itertools
+        g = itertools.cycle(pal)
+        for i in range(len(chunk)):
+            colors.append(next(g))
+
+        return colors
+
+    def _set_and_get(self, data, prefix, attr, val, content):
+        """Set a new attr and then get it to fill the self.data dict.
+
+        Keep track of the attributes created.
+
+        Args:
+            data (dict): where to store the new attribute content
+            attr (list): where to store the new attribute names
+            val (string): name of the new attribute
+            content (obj): content of the new attribute
+        """
+        setattr(self, prefix + val, content)
+        data[prefix + val] = getattr(self, prefix + val)
+        attr.append(prefix + val)
+
+    def set_and_get(self, prefix, val, content):
+        """Set a new attr and then get it to fill the self.data dict.
+
+        Keep track of the attributes created.
+
+        Args:
+            prefix (str): prefix of the new attribute
+            val (string): name of the new attribute
+            content (obj): content of the new attribute
+        """
+        self._set_and_get(self.data, prefix, self.attr, val, content)
+
+    @property
+    def palette(self):
+        """Build a color list just cycling through a defined palette.
+
+        """
+        if not self._palette:
+            self._palette = self._set_colors(self.groups)
+
+        return self._palette
+
+    def make_scatter(self, source, x, y, markertype, color, line_color=None,
```

Comment:
This should definitely go in a `utils.py` or something but that can wait until later.